Flink Storm compatibility

mjsax opened the pull request:
First version of a Storm compatibility layer for Flink. It enables the execution of Storm Topologies with Flink. Additionally, Storm Spouts and Bolts can be used within regular Flink streaming programs.

Current limitations:
  - only Storm's default output stream is supported
  - only shuffle and fields-grouping supported
  - no meta-data headling (ie, Configuration and TopologyContext) for Spouts and Bolts

mbalassi commented:
Hey Matthias,

Thanks for opening the pull request as discussed on the mailing list. Could you please provide some description here also?

Please do another update on your PR to better match our coding guidelines:

1. The first six commits are not needed. Th first one is already merged, than you have 5 huge merge commits, that we are trying to avoid.
2. Some of your commit messages need a bit more polishing, the "added raw output type support" and "added API compatibility classes" do not describe your work when merged to the Flink code. I would also suggest to add a [streaming] tag for your commits.
3. Please squash 605de53 to 1519d36, becuase it is just a minor fix for that.

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkClient.java:
```diff
@@ -0,0 +1,362 @@
+ * Nimbus.Client} at once, to interact with Flink's JobManager instead of Storm's Nimbus.
+ * 
+ * @author mjsax
+ */
```
No author tags, please.

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkClient.java:
```diff
@@ -0,0 +1,362 @@
+	/**
+	 * Configuration key to specify the location of the file json-simple-1.1.jar (for com.googlecode.json-simple).
+	 */
+	public final static String FLINK_STROM_JSON_SIMPLE_JAR_LOCATION = "flink.storm.jar-location.json-simple";
```
Do we need to hard code these?

mjsax replied:
No really. It was the least intrusive way to make it work....

In order to be able to execute a Storm topology on Flink, both jars are needed. The build configuration could be changed to include both in the final Flink jar. Because I did not want to change the build without feedback, I added both parameter keys to enalbe the user to specifiy the location for both jars manually.

Right now, both parameters are used within StormWordCountRemote (in a ugly hardcoded way). This could also be removed, if both jars would be included in the final Flink jar.

mjsax replied:
I fixed this issue. (see flink-dist/pom.xml, FlinkClient.java, and StormWordCountRemote.java)
In order to get a clean module structure, I extracted the examples from flink-storm-compatiblitly into a own module flink-storm-examples.

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkClient.java:
```diff
@@ -0,0 +1,362 @@
+			
+		} catch(final Exception e) {
+			e.printStackTrace();
+		}
```
This happens on the client side, but I would still log it instead of printing stack trace.

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkClient.java:
```diff
@@ -0,0 +1,362 @@
+	
+	// public String getNimbusConf() {
+	// return null;
+	// }
```
This is a bit too much comment here to merge, but we should definitely discuss these.

mjsax replied:
I would like to keep it or to change it to throw UnsupportedOperationException(). This ensures, that the user is aware what operations are supported (or not supported) by the compability layer.

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkLocalCluster.java:
```diff
@@ -0,0 +1,135 @@
+ * @author mjsax
+ */
+public class FlinkLocalCluster {
+	
```
The class looks like a stub itself :)

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkOutputFieldsDeclarer.java:
```diff
@@ -0,0 +1,173 @@
+	 * {@link DummyComparable} is a {@link Comparable} helper class that is used to get the correct
+	 * {@link TypeInformation} from {@link TypeExtractor} within {@link #getOutputType()}. If key fields are not
+	 * comparable, Flink cannot use them and will throw an exception.
+	 * 
```
A not word is missing from the javadoc. If keys are not comparable then an exception is thrown. Maybe a rename would also be nice.

mjsax replied:
As the comment says: "If key fields are NOT comparable, Flink cannot use them and will throw an exception."

Any suggestion for a better name?

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkSubmitter.java:
```diff
@@ -0,0 +1,258 @@
+		
+		if(localNimbus != null) {
+			// TODO: check what is the difference about localNimbus vs LocalCluster
+			// LOG.info("Submitting topology " + name + " in local mode");
```
Let us resolve this, before merging.

mjsax replied:
I ask at the dev-storm mailing list and got the following answer:

"LocalNimbus is something that is not really used.  It provides a way for the StormSubmitter to be used when submitting a topology to a local mode cluster.  No one uses it and in the past it didn't work, not sure if it works now or not.  It is probably best to just not use it unless you really need to abstract out local mode from the topology code used to submit a topology.
 - Bobby"

We can ignore this case, and I will remove the code.

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkTopology.java:
```diff
@@ -0,0 +1,113 @@
+		super.setParallelism(1);
+		this.stormTopology = stormTopology;
+	}
+	
```
Does it work for higher parallelism?

mjsax replied:
Storm uses default parallelism of one, thus, we set it to one here. We do not want to break potential user assumptions about default dop.

It also works for different dop values; the user has to specify non-default dop when adding Spouts/Bolts to the topology.

mbalassi replied:
Ok, cool. Then just let us put a comment there, that this value is to mirror Storm behaviour.

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkTopologyBuilder.java:
```diff
@@ -0,0 +1,294 @@
+	 */
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	public FlinkTopology createTopology() {
+		final StormTopology stormTopolgoy = this.stormBuilder.createTopology();
```
Nice! :+1:

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkTopologyContext.java:
```diff
@@ -0,0 +1,221 @@
+		// if(bolt != null) {
+		// return bolt.get_common();
+		// }
+		// final SpoutSpec spoutSpec = this.topology.get_spouts().get(componentId);
```
Couple of things seems a bit like overkill here. :)

mjsax replied:
I prefer throwing exceptions, such that the user is aware what Storm features are not supported. However, this class is in an pre-alpha state... It might be necessary to add some more stuff here, to get meta-data handling right (ie, hand an correct TopooogyContext object to Spouts/Bolts within .open()/.prepare().

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/examples/BoltTokenizerWordCount.java:
```diff
@@ -0,0 +1,127 @@
+ * </ul>
+ */
+public class BoltTokenizerWordCount {
+	
```
Nice! :+1:

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/wrappers/StormSpoutWrapper.java:
```diff
@@ -0,0 +1,88 @@
+
+
+
+
```
You use too much spacing in this file.

mjsax replied:
Why? Can it be a problem?

mbalassi replied:
It is just better to have a somewhat unified look of our code. No semantic problem of course.

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/wrappers/StormTuple.java:
```diff
@@ -0,0 +1,251 @@
+ * @author mjsax
+ */
+class StormTuple<IN> implements Tuple {
+	/**
```
Looking at this file I am really thankful for the Flink Tuple type. Can't we support the operations you marked unsupported?

mjsax replied:
Supporting named attributes is possible (and I guess, not too hard). Should it be part in the first pull request? Or can we add it later on?

mbalassi commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/wrappers/StormWrapperSetupHelper.java:
```diff
@@ -0,0 +1,123 @@
+		
+		return new FlinkTopologyContext(new StormTopology(spoutSpecs, bolts, null), null, taskToComponents, null, null,
+			null, null, null, taskId, null, null, null, null, null, null, null);
+	}
```
That constructor needs a revision. :)

mjsax replied:
TopologyContext is not supported so far. This is only a pre-alpha version to make it compile and run. Added TopologyContext (in a nice way) need some more work. Not sure If we need it for the first pull request?

mjsax commented:
I fixed the rebase problem to remove the unnecessary commits. I also squashed the commits you mentioned, removed author tags, and improved commit messages.

mbalassi commented:
Thanks, let me do another pass over the code later this week.

mbalassi commented:
Thanks for the updates.

Although the examples are self-explanatory could you please add a readme on a minimal use case and explaining the features that you currently cover, please?

rmetzger commented on flink-staging/flink-streaming/flink-storm-examples/src/main/java/org/apache/flink/stormcompatibility/wordcount/stormoperators/StormFileSpout.java:
```diff
@@ -0,0 +1,84 @@
+			try {
+				this.reader.close();
+			} catch(final IOException e) {
+				e.printStackTrace();
```
These exceptions should be re-thrown as a runtime exception so that the system can properly fail (instead of logging a stacktrace to sysout

rmetzger commented on flink-staging/flink-streaming/flink-storm-examples/src/main/java/org/apache/flink/stormcompatibility/wordcount/stormoperators/StormBoltFileSink.java:
```diff
@@ -0,0 +1,77 @@
+		try {
+			this.writer = new BufferedWriter(new FileWriter(this.path));
+		} catch(final IOException e) {
+			e.printStackTrace();
```
Again, the exception handling in this file is not correct

rmetzger commented on flink-staging/flink-streaming/flink-storm-examples/src/main/java/org/apache/flink/stormcompatibility/wordcount/StormWordCountLocal.java:
```diff
@@ -0,0 +1,80 @@
+		final FlinkLocalCluster cluster = FlinkLocalCluster.getLocalCluster();
+		cluster.submitTopology(topologyId, null, builder.createTopology());
+		
+		Utils.sleep(5 * 1000);
```
Why is it necessary to sleep in the examples? Isn't there a way to submit the topology in a blocking way?

mjsax replied:
Storm submitts topologies always in a non-blocking way. This behavior is mimiced in the compatibility layer.

rmetzger commented on flink-staging/flink-streaming/flink-storm-compatibility/src/test/java/org/apache/flink/stormcompatibility/wrappers/StormWrapperSetupHelperTest.java:
```diff
@@ -0,0 +1,163 @@
+	public void prepareTest() {
+		final long seed = System.currentTimeMillis();
+		this.r = new Random(seed);
+		System.out.println("Test seed: " + seed);
```
Can you use our logging framework for these messages?
We set different logging levels for different environments (mvn locally, travis, ide)

mjsax replied:
Sure. What logging level should I use? I want to make sure to see the seed anytime a build fails to be able to reproduce the test.

rmetzger replied:
I'd suggest "info"

rmetzger commented on flink-staging/flink-streaming/flink-storm-compatibility/src/test/java/org/apache/flink/stormcompatibility/wrappers/StormTupleTest.java:
```diff
@@ -0,0 +1,367 @@
+	public void prepare() {
+		this.seed = System.currentTimeMillis();
+		this.r = new Random(this.seed);
+		System.out.println("Test seed: " + this.seed);
```
use logging framework

rmetzger commented on flink-staging/flink-streaming/flink-storm-compatibility/src/test/java/org/apache/flink/stormcompatibility/wrappers/StormFiniteSpoutWrapperTest.java:
```diff
@@ -0,0 +1,124 @@
+		final long seed = System.currentTimeMillis();
+		this.r = new Random(seed);
+		System.out.println("Test seed: " + seed);
+	}
```
maybe it would make sense to create a test base for the storm compat layer to avoid duplicate code.

rmetzger commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkClient.java:
```diff
@@ -0,0 +1,321 @@
+	 *            The jobmanager's rpc port.
+	 * @param timeout
+	 */
+	public FlinkClient(final Map<?, ?> conf, final String host, final int port, final Integer timeout) {
```
The `conf` argument is unsued

rmetzger replied:
Why is the Map not typed to Map<String, String>?

mjsax replied:
FlinkClient offers the unified interface as NimbusClient and Nimbus.Client (in the compatibility layer, it does not make sense to destingues both). Thus, the constructor offers the same parameters (even in not used right now). The same holds for Map -- Storm does not use any generics.

rmetzger commented on flink-staging/flink-streaming/flink-storm-compatibility/src/main/java/org/apache/flink/stormcompatibility/api/FlinkClient.java:
```diff
@@ -0,0 +1,321 @@
+				try {
+					Thread.sleep(1000 * options.get_wait_secs());
+				} catch(final InterruptedException e) {
+					e.printStackTrace();
```
rethrow as runtime e.

mjsax commented:
Hi, I addressed all your comments. The latest build was successful, too. Please let me know if there is anything else to improve the pull request.

StephanEwen commented:
Anything blocking this pull request, or can it be merged?

mjsax commented:
Hi, I rebased to the current master (and new streaming API) as discussed and cleaned up everything. Unfortunately, I was not able to cherry-pick the last commits:
 - @szape commits: https://github.com/mbalassi/flink/commit/ed189aac5de79f1af50efcbb0a1d9defcc582d1a
 - @mbalassi https://github.com/mbalassi/flink/commit/9704649267dd29291bb14f223f17a18fc2c26cd0
 - @mbalassi https://github.com/mbalassi/flink/commit/3837880544149361a74a2b9b915346e41c46dbac

@szape commit is nor crucial and can be applied right now (please per pull request to mjsax/flink-storm-compatibiltiy) or later on.
Both commits from @mbalassi are critical and I re-did the changes manually (see last commit of this PR). Of course, I accounted it to @mbalassi as an author.

From my point of view, this PR is ready to be merged (hope that Travis will be green).

mjsax commented:
Any news on this PR?

mbalassi commented:
Hey @mjsax, thanks for the cleanup. I was focused on the release during the week, I'm doing the merge this evening.

mbalassi commented:
Did a final pass and a rebase, code now looks good. There is one little issue with the tests that I need to check again tomorrow morning then I'm merging it. :+1:

mbalassi commented:
Turned logging off for tests by default and did minor javadoc and typo cleanup. We should address the job cancelling throwing expected exceptions soonish. But I think we can merge it to contrib in the current state.

mjsax commented:
Nice :)
Why did you turn off logging? I have some randomized tests and if they fail, the test error cannot be reproduces without the seed value that goes to the log...
About canceling jobs: I have already an open pull request for it: https://github.com/apache/flink/pull/750
